packages/www/hooks/use-api.tsx

```diff
@@ -840,6 +854,95 @@ const makeContext = (state: ApiState, setState) => {
       return asset;
     },
 
+    async uploadAssets(files: File[]): Promise<void> {
+      const requestAssetUpload = async (
+        params
+      ): Promise<{ tusEndpoint: string }> => {
+        const [res, assetUpload] = await context.fetch(
+          `/asset/request-upload`,
+          {
+            method: "POST",
+            body: JSON.stringify(params),
+            headers: {
+              "content-type": "application/json",
+            },
+          }
+        );
+
+        if (!res.ok) {
+          throw new Error(assetUpload.errors.join(", "));
+        }
+        return assetUpload;
+      };
+
+      const updateStateWithProgressOrError = (
+        file: File,
+        progress: number,
+        completed: boolean,
+        error?: Error
+      ) => {
+        setState((state) => ({
+          ...state,
+          currentFileUploads: {
+            ...state.currentFileUploads,
+            [file.name]: {
+              file,
+              progress,
+              error,
+              completed:
+                state?.currentFileUploads?.[file.name]?.completed ||
+                Boolean(completed),
+            },
+          },
+        }));
+      };
+
+      const getTusUpload = (file: File, tusEndpoint?: string) =>
+        new tus.Upload(file, {
+          endpoint: tusEndpoint ?? undefined, // URL from `tusEndpoint` field in the `/request-upload` response
+          metadata: {
+            filetype: file.type,
+          },
+          uploadSize: file.size,
+          onError(err) {
+            updateStateWithProgressOrError(file, 0, false, err);
+          },
+          onProgress(bytesUploaded, bytesTotal) {
+            const percentage = bytesUploaded / bytesTotal;
+            updateStateWithProgressOrError(file, percentage, false);
+          },
+          onSuccess() {
+            updateStateWithProgressOrError(file, 1, true);
+          },
+        });
+
+      for (const file of files) {
+        try {
+          updateStateWithProgressOrError(file, 0, false);
+
+          const uploadWithoutUrl = getTusUpload(file);
+          const previousUploads = await uploadWithoutUrl.findPreviousUploads();
```

Comment:
I actually think you won't be able to find previous uploads without the URLs here, since the fingerprint key used to find the upload actually includes the `endpoint` opt in the upload: https://github.com/tus/tus-js-client/blob/6f723f07b016cd374d4fe65140a2794e8ca7ca56/lib/browser/fileSignature.js#L23

Did you manage to observe something different than that? Like uploads being restored in this initial flow.

So to get retry-able uploads even if the browser refreshes, I guess you'd need to manage a separate local storage entry with current uploads and their corresponding `tusEndpoints`. I don't think it's necessary to include that in this PR (especially since if you refresh you're gonna lose all the `File` references anyway), but if you do so lmk and I can help review it as well :)

Reply:
Ahh I'm just seeing this comment. This makes sense, and explains the behavior I outlined in my other comment. I think we should file a bug for state management on the frontend to track `file->uploadEndpoint` and then reuse that upload endpoint if it exists. What do you think?

Reply:
Added an issue here: https://github.com/livepeer/studio/issues/1203

Reply:
Yeah nice, I agree! We'd need to store the `uploadEndpoint` and also the file name somehow, so we only try to resume if it is the same file being uploaded (or even prompt the user like "hey you were uploading X, wanna continue? pick it on your filesystem again")